### PR TITLE
feat: added SQL error messages and functions

### PR DESCRIPTION
**Pull request by Xhoenix**
Source: https://portswigger.net/web-security/sql-injection/cheat-sheet#extracting-data-via-visible-error-messages

### CODE REVIEW

**Review by theseion — CHANGES_REQUESTED** [review]
Great additions, thanks!

Please add at least one test for each of the new entries and document the source for each entry (the page you linked to does not list the error messages you added, for example).

**Comment by theseion** [comment]
I don't see any changes to test files in your PR.

Screenshot 2023-10-16 at 20 25 49

**Comment by Xhoenix** [comment]
@theseion is a test really necessary? I just checked @azurit has created the tests with something like this:-

https://github.com/coreruleset/coreruleset/blob/249a0b7aa4cb18f50577f47bdb38d5836044524b/tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml#L25-L27

As we're checking the response body for the desired error messages, it's just a simple copy/paste of the rules created by @azurit

**Comment by theseion** [comment]
In my opinion yes, the tests are necessary. We have too little test coverage. I know that it's a little bothersome but doing this kind of work pays off in the long run. Not only that, we will be able to reuse tests for things like performance measurements.

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml [hunk 1]
```diff
@@ -25,3 +25,23 @@ tests:
...
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
+            version: "HTTP/1.0"
+            uri: "/anything"
+            data: "[match sql-errors.data]the used select statements have different number of columns[/match]: ERROR:  invalid input syntax for type integer"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml:45]:
```suggestion
            data: "[match sql-errors.data]the used select statements have different number of columns[/match]: ERROR:  invalid input syntax for integer"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ---
 meta:
-  author: "azurit"
+  author: "azurit and Xhoenix"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml:3]:
```suggestion
  author: "azurit, Xhoenix"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ---
 meta:
-  author: "azurit"
+  author: "azurit and Xhoenix"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml:3]:
```suggestion
  author: "azurit, Xhoenix"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ---
 meta:
-  author: "azurit"
+  author: "azurit and Xhoenix"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml:3]:
```suggestion
  author: "azurit, Xhoenix"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml [hunk 5]
```diff
@@ -25,3 +25,24 @@ tests:
...
+            headers:
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
+            version: "HTTP/1.0"
+            uri: "/anything"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml:45]:
```suggestion
            uri: "/get"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml [hunk 6]
```diff
@@ -25,3 +25,23 @@ tests:
...
+            headers:
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
+            version: "HTTP/1.0"
+            uri: "/anything"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml:44]:
```suggestion
            uri: "/get"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml [hunk 7]
```diff
@@ -25,3 +25,23 @@ tests:
...
+            headers:
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
+            version: "HTTP/1.0"
+            uri: "/anything"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml:44]:
```suggestion
            uri: "/get"
```

**Review by theseion — CHANGES_REQUESTED** [review]
That looks pretty good, thanks. Just a bit of clean up and a test fix.

**Comment by Xhoenix** [comment]
I'll make it a single commit. :)

**Comment by theseion** [comment]
The change from `/anything` to `/get` causes the tests to fail. The data in the `GET` body will not be reflected by the backend, because `GET`s are not supposed to have a body. Hence, the rule will not detect anything in the response body. Changing the request method to `POST` solves this issue.

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml [hunk 8]
```diff
@@ -25,3 +25,24 @@ tests:
...
+            headers:
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
+            version: "HTTP/1.0"
+            uri: "/get"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml:45]:
```suggestion
            uri: "/post"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml [hunk 9]
```diff
@@ -25,3 +25,24 @@ tests:
...
+            dest_addr: "127.0.0.1"
+            port: 80
+            headers:
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951220.yaml:43]:
```suggestion
            method: "POST"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml [hunk 10]
```diff
@@ -25,3 +25,23 @@ tests:
...
+            dest_addr: "127.0.0.1"
+            port: 80
+            headers:
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml:42]:
```suggestion
            method: "POST"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml [hunk 11]
```diff
@@ -25,3 +25,23 @@ tests:
...
+            version: "HTTP/1.0"
+            uri: "/get"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951230.yaml:44]:
```suggestion
            uri: "/post"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml [hunk 12]
```diff
@@ -25,3 +25,23 @@ tests:
...
+            dest_addr: "127.0.0.1"
+            port: 80
+            headers:
+              Host: "localhost"
+              User-Agent: "OWASP CRS test agent"
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+              Accept-Encoding: "gzip,deflate"
+              Accept-Language: "en-us,en;q=0.5"
+              Content-Type: "text/plain"
+            method: "GET"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml:42]:
```suggestion
            method: "POST"
```

File: tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml [hunk 13]
```diff
@@ -25,3 +25,23 @@ tests:
...
+            version: "HTTP/1.0"
+            uri: "/get"
```
Comment (theseion) [inline tests/regression/tests/RESPONSE-951-DATA-LEAKAGES-SQL/951240.yaml:44]:
```suggestion
            uri: "/post"
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Comment by Xhoenix** [comment]
Guys, LGTM.

**Review by theseion — APPROVED** [review]

**Comment by theseion** [comment]
Thanks!

**Pull request merged**